Make tele intercept --docker-run play well with tele connect --docker

thallgren opened the pull request:
## Description

A lot of improvements were made in how we run the Telepresence daemon in a Docker container when the [Telepresence Intercept Specification](https://www.getambassador.io/docs/telepresence/latest/reference/intercepts/specs#configuring-intercept-using-specifications) was introduced. A major one was that an intercept handler that runs in docker, now can talk to the daemon without using the host network. Another that volumes could be mounted using the [Telemount plugin](https://github.com/datawire/docker-volume-telemount), removing the need for FUSE software like macFUSE and Winfsp. We didn't touch the `telepresence intercept --docker-run` flag at that time though, which lead to weird results when combining it with a daemon started with `telepresence connect --docker`. This PR rectifies that.

If the daemon is started in a container with `telepresence connect --docker` (or implicitly by passing `--docker` directly to `telepresence intercept`), and the `--docker-run` is passed to the intercept command, then Telepresence will automatically connect the dots between the intercept handler and the daemon container. It will also automatically mount the remote filesystems that are made available by the traffic-agents SFTP server.

Two new flags are also added in this PR. The `--docker-build <docker context>` and the repeatable `docker-build-opt key=value`, to enable container's to be build on the fly by the intercept command. The `--docker-build` implies `--docker-run`.

The PR also contains some other improvements (details in the commits).

## Checklist

 - [x] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to add any docs changes required for my change (including release notes).
 - [x] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.
 - [ ] I updated `TELEMETRY.md` if I added, changed, or removed a metric name.
 - [ ] Once my PR is ready to have integration tests ran, I posted the PR in #telepresence-dev in the datawire-oss slack so that the "ok to test" label can be applied.

Commit:
Prevent panic when running command with intercept and quitting

The client would sometimes panic when running an intercept in one
terminal that also started a command (a `--docker-run` for instance) and
then doing a `telepresence quit` in another.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Clean up ambiguous logic regarding `--mount` flag in intercept.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move docker-run related functions into a docker_run.go file.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move CreateRequest in state.go to other public state functions.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Convert to errcat.NoDaemonLogs when calling intercept.CreateRequest

Instead of doing it inconsistently inside of that function.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Cache connector Status and intercept Info in intercept.status struct

They are needed in future commits.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure proper stop of intercept handler container.

The container is stopped using `docker stop` instead of just killing
the `docker run` with a signal, regardless of if the stop is triggered
by a signal, by a `telepresence leave`, or a `telepresence quit`.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix problem with the DNS recursion check.

Adds a short timeout and a repeat to the DNS recursion check so that
it doesn't need to timeout in the cluster. The rationale behind this is
that if the DNS indeed is recursive, then the cluster's DNS server must
have access to the network where Telepresence DNS server is present on
on the client, which means that the cluster must be on the client and
the DNS should therefor be very fast.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add debug logging to proc.Start.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix combined `tele connect --docker` and `tele intercept --docker-run`

When the Telepresence daemon runs in a docker container (as the result
of passing the `--docker` flag to `telepresence connect`), and an
intercept starts an intercept handler with `--docker-run`, then the
handler container must utilize the network of the daemon container, and
use the `telemount` volume driver plugin for volume mounts.

This commit adds everything that is needed to ensure that those dots
are connected between the handler and the daemon.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add intercept flag `--docker-build`.

The `--docker-build <docker context>` is similar to `--docker-run` but
instead of running an existing image, it will first build the image
from a Docker context (URL or DIR).

Options to the `--docker-build` can be passed using repeated
`--docker-build-opt key=value` flags.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Bump github.com/docker/docker to v20.10.24+incompatible

To get rid of dependabot alerts.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix panic when no tunVif device is present.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make recursion check work again on macOS.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

knlambert commented on DEPENDENCIES.md:
```diff
@@ -30,7 +30,7 @@ following Free and Open Source software:
     github.com/docker/cli                                                        v20.10.22+incompatible                Apache License 2.0
     github.com/docker/distribution                                               v2.8.1+incompatible                   Apache License 2.0
-    github.com/docker/docker                                                     v20.10.22+incompatible                Apache License 2.0
+    github.com/docker/docker                                                     v20.10.24+incompatible                Apache License 2.0
```
I think you forgot to update the changelog

knlambert commented on pkg/client/cli/intercept/docker_run.go:
```diff
@@ -0,0 +1,218 @@
+	//
+	// telepresence intercept hello --docker-build ./some/path -- -it --add-host foo IMAGE --port 8080
+	if (idx < 0 || imageName != "IMAGE") && len(s.Cmdline) > 0 {
+		return errcat.User.New(`` +
```
Empty `` ?

thallgren replied:
Yeah, because otherwise the formatting of the lines that follow gets messed up.

Commit:
Add a timeout to the scout-disable check.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix integration_test glitch that caused clean-ups to be forgotten.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add tests for docker-run with container based daemon.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix recursionCheck on Windows.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Update CHANGELOG.md

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Update text for the `--port` intercept flag.

It behaves differently with `--docker-run` only when the container
doesn't run in docker.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Update the golangci-lint dependencies.

Because `make lint` causes OOM in the telepresence2-proprietary repo.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure windows searchList contains unique entries.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make the resolveFile struct "unix" instead of out of "darwin"-only.

No need to maintain two parses for this kind of file.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Replace search-path that isn't intended for the cluster DNS resolver.

This applies only to the overriding resolver that we use in Docker and
on Linux where systemd.resolved isn't configured.

Instead of adding a search path declared in a resolv.conf file to the
exclude-suffix list, this commit ensures that such paths are is stripped
off before the query is subjected to Telepresence search-path and sent
to the cluster.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Don't report errors from fallback resolver when there's a local hit

This commit addresses the problem where the Telepresence resolver finds
an entry for an A record, but not for an AAAA record, and the AAAA
query therefore is sent to the fallback resolver. In such situations,
the NXNAME error from the AAAA query must be replaced by an EMPTY
result, indicating that the A record is sufficient. The same is true
for the opposite where an AAAA record is found but the A is sent to
the fallback.

Fallback servers are only used by the overriding resolver.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix lint error.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Give recursion check more retries and longer timeout on Windows.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Don't error when leave doesn't find an intercept.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
